pkg/openstack/loadbalancer.go

```diff
@@ -2220,47 +2222,102 @@ func (lbaas *LbaasV2) ensureAndUpdateOctaviaSecurityGroup(clusterName string, ap
+				Direction:      rules.DirIngress,
+				Protocol:       rules.ProtocolTCP,
+				EtherType:      etherType,
+				RemoteIPPrefix: subnet.CIDR,
+				SecGroupID:     lbSecGroupID,
+				PortRangeMin:   int(apiService.Spec.HealthCheckNodePort),
+				PortRangeMax:   int(apiService.Spec.HealthCheckNodePort),
+			},
 		)
-		if err != nil {
-			return fmt.Errorf(
-				"failed to apply security rule for health check node port, %w",
-				err)
-		}
 	}
 
-	// ensure rules for node security group
 	for _, port := range ports {
 		if port.NodePort == 0 { // It's 0 when AllocateLoadBalancerNodePorts=False
 			continue
 		}
-		err = lbaas.ensureSecurityRule(
-			rules.DirIngress,
-			rules.RuleProtocol(port.Protocol),
-			etherType,
-			subnet.CIDR,
-			lbSecGroupID,
-			int(port.NodePort),
-			int(port.NodePort),
+		wantedRules = append(wantedRules,
+			rules.CreateOpts{
+				Direction:      rules.DirIngress,
+				Protocol:       rules.ProtocolTCP,
+				EtherType:      etherType,
+				RemoteIPPrefix: subnet.CIDR,
+				SecGroupID:     lbSecGroupID,
+				PortRangeMin:   int(port.NodePort),
+				PortRangeMax:   int(port.NodePort),
+			},
 		)
+	}
+
+	toCreate := make([]rules.CreateOpts, 0, len(ports)+1)         // Max is number of ports+1
+	toDelete := make([]rules.SecGroupRule, 0, len(existingRules)) // Max will be all the existing rules
+	// Surely this can be done in a more efficient way. Is it worth optimizing if most of
+	// the time we'll deal with just 1 or 2 elements in each array? I doubt it.
+	for _, rule := range existingRules {
+		found := false
+		for _, opts := range wantedRules {
+			if compareSecurityGroupRuleAndCreateOpts(rule, opts) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			// in existingRules but not in wantedRules, delete
+			toDelete = append(toDelete, rule)
+		}
+	}
+	for _, opts := range wantedRules {
+		found := false
+		for _, rule := range existingRules {
+			if compareSecurityGroupRuleAndCreateOpts(rule, opts) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			// in wantedRules but not in exisitngRules, create
+			toCreate = append(toCreate, opts)
+		}
+	}
+
+	// create new rules
+	for _, opts := range toCreate {
+		err := lbaas.ensureSecurityRule(opts)
 		if err != nil {
-			return fmt.Errorf(
-				"failed to apply security rule for port %d, %w",
-				port.NodePort, err)
+			return fmt.Errorf("failed to apply security rule (%v), %w", opts, err)
 		}
+	}
 
-		if err := applyNodeSecurityGroupIDForLB(lbaas.compute, lbaas.network, nodes, lbSecGroupID); err != nil {
-			return err
+	// delete unneeded rules
+	for _, existingRule := range toDelete {
+		klog.Infof("Deleting rule %s from security group %s (%s)", existingRule.ID, existingRule.SecGroupID, lbSecGroupName)
+		mc := metrics.NewMetricContext("security_group_rule", "delete")
+		err := rules.Delete(lbaas.network, existingRule.ID).ExtractErr()
+		if err != nil && cpoerrors.IsNotFound(err) {
+			// ignore 404
+			return mc.ObserveRequest(nil)
```

Comment:
For this and the 419 above, actually hitting this should only be possible if we're making concurrent modifications elsewhere. I think we should log them both at a minimum, and possibly also fail. If we fail we'll just retry, presumably without conflicts.

Reply:
419?

Reply:
Oh, 409. Okay, makes sense. I'll stick with a warning. If rule exists already or is gone already, we're fine.